### ISSUE TITLE
`TypeIs[]` incorrectly narrows tuple with known length

### ISSUE THREAD

**Issue by InSyncWithFoo** [microsoft/pyright#8554]
Reproducible example ([playground](https://pyright-play.net/?pyrightVersion=1.1.373&pythonVersion=3.14&strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgJIDOAsAFA0Amxwm5A%2BjAK4IA2xLYwLcjBCo05ABQA3AFyxOPANqp8AHyhCQAGigA6PQF0AlFAC0APkIkKCjt2IKN2vTv37ZzmvUZQA7sSQgdCwAxmAowQCGMMQoUcTiESBisraKylBqjpgoqurCRqYWGtI0UGVQIMSSxBFcbCQJSeTG5WUAxFChIJXBMLJydko5GXla2bka%2Bp7UrUhMKGD4SKypvPyCwqISiWKGJTOt5ZXVtfUI8TvN7Z24PX1lq0MTwtrpmfmlh2UgEcvElucAKLdXDTVrHGp1eDnRq7L4dYgAD3OvWIdBS8nsWUmny%2BePxrQ6EV67Fq-SgAB5yOwAEbBLgRcjkKD8AY8KARFB0NnEMxg8qI7RwKAAXg5TVxFWIHBAKCgiKgAGooMKCVAEcjiKj0VBgKhiJK1UbrsSOGTygAlaXsWWwEjaABEbygAFUUABrBY%2BOXvEAO17MhCJGBIWpcYXsD1elBAA)):

```python
from typing import TypeIs

def is_tuple_of_strings(v: tuple[int | str, ...]) -> TypeIs[tuple[str, ...]]: ...

def weird_concatenate(args: tuple[int | str, int | str]) -> str:
    reveal_type(args)      # correct:  tuple[int | str, int | str]

    if not is_tuple_of_strings(args):
        reveal_type(args)  # correct:  tuple[int | str, int | str]
        raise TypeError

    reveal_type(args)      # expected: tuple[str, str]
                           # actual:   <subclass of tuple and tuple>

    x, y = args
    return x + y           # expected: fine
                           # actual:   Return type, "int | Unknown | str", is partially unknown
```

Yet it is able to do a better job when `args`'s length is unknown ([playground](https://pyright-play.net/?pyrightVersion=1.1.373&pythonVersion=3.14&strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgJIDOAsAFA0Amxwm5A%2BjAK4IA2xLYwLcjBCo05ABQA3AFyxOPANqp8AHyhCQAGigA6PQF0AlFAC0APkIkKCjt2IKN2vTv37ZzmvUZQA7sSQgdCwAxmAowQCGMMQoUcTiESBisraKylBqjroGxubqwtI0UMVQIMSSxBFcbCQJSeTGJcUAxFChIGXBMLJydkooqvla2S6e1E1ITChg%2BEisqbz8gsKiEolihoXjTSVlFVU1CPHrDS1tuJ3dxQv9g1nO%2BkU7xSARc8SWRwCiHbhjTXtKtV4Ec6htnq12pcejd7gZ-iUAB7aOBQAC8UBOTxexA4IBQUERUAA1FBUc8mpCLsQuj1gKhiDQgA)):

```python
def weird_concatenate(args: tuple[int | str, ...]) -> str:
    reveal_type(args)      # correct:  tuple[int | str, ...]

    if not is_tuple_of_strings(args):
        reveal_type(args)  # correct:  tuple[int | str, ...]
        raise TypeError

    reveal_type(args)      # correct:  tuple[str, ...]

    x, y = args
    return x + y           # correct:  fine
```

A similar error happens when a `TypeVar` is used ([playground](https://pyright-play.net/?pyrightVersion=1.1.373&pythonVersion=3.14&strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgJIDOAsAFA0Amxwm5A%2BjAK4IA2xLYwLcjBCo05ABQA3AFyxOPANqp8AHyhCQAGigA6PQF0AlFAC0APkIkKCjt2IKN2vTv37ZzmvUZQA7sSQgdCwAxmAowQCGMMQoUfYEsuLK2hqG%2BuIRIGKytooETgbG5urC0jRQFVAgxJLEEVxsJBlZ5MaVFQDEUKEg1cEwsnJ2Cvm6Bp7U7UhMKGD4SKy5vPyCwqISmWKGZZPtldW19Y0IxM1bnVDEAB4n-cR0OfL2yWMu5Xsfn19dEf3s9YMhnkCm9dh8QBEFsRLCcAKK9XATdoHOoNeAnM6tD5da63aIPIH2RyvfTvL7k75QX4cAGVAA85HYACNglwIuRyFB%2BISqSg6ISzEjKldtHAoABeKktMlVYgcEAoKBXKAAaigYopOJuxDuBOAqGIMopxou1P%2BXEBRBOUAARMoAFRQNQae025iYcKQBBRJBMni%2BJAwAAWsBIto0NpoQA)), [regardless of length](https://pyright-play.net/?pyrightVersion=1.1.373&pythonVersion=3.14&strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoAqiApgJIDOAsAFA0Amxwm5A%2BjAK4IA2xLYwLcjBCo05ABQA3AFyxOPANqp8AHyhCQAGigA6PQF0AlFAC0APkIkKCjt2IKN2vTv37ZzmvUZQA7sSQgdCwAxmAowQCGMMQoUfYEsuLK2hqG%2BuIRIGKytooE2gRGphYa0jRQFVAgxJLEEVxsJBlZ5MaVFQDEUKEg1cEwsnJ2Cvm6Bp7U7UhMKGD4SKy5vPyCwqISmWKGZZPtldW19Y0IxM1bnVDEAB4n-cR0OfL2yWMu5Xsfn19dEf3s9YMhnknONdh8QBEFsRLCcAKK9XATdoHOoNeAnM6tD5da63aIPIH2RyvfTvL7k75QX4cAGVAA85HYACNglwIuRyFB%2BISqSg6ISzEjKldtHAoABeKktMlVYgcEAoKBXKAAaigYopOJuxDuBOAqGIMopxou1P%2BXEBRBOUAARMoAFRQNQae025iYcKQBBRJBMni%2BJAwAAWsBIto0NpoQA):

```python
def weird_concatenate[T: (int, str)](args: tuple[T, ...]) -> str:
    reveal_type(args)      # correct:  tuple[T, ...]

    if not is_tuple_of_strings(args):
        reveal_type(args)  # expected: tuple[int, ...]
                           # actual:   tuple[T, ...]
        raise TypeError

    reveal_type(args)      # expected: tuple[str, ...]
                           # actual:   <subclass of tuple and tuple>

    x, y = args
    return x + y           # expected: fine
                           # actual:   Type "int* | str*" is incompatible with type "str"
```

**Comment by erictraut** [comment]
Thanks for the bug report.

I view this as two separate issues. First, pyright is currently producing incorrect narrowed types in this case. The resulting types are overly narrow or just plain wrong. I consider that a bug, and it should be fixed. For comparison, mypy is also producing incorrect narrowed types in this case. I've filed [a bug](https://github.com/python/mypy/issues/17599).

I've prepared a fix for the bug. With this fix in place, pyright no longer produces incorrect (too narrow or nonsensical) narrowed types, but it doesn't produce the optimal types that you are hoping to see here. Instead, it refrains from narrowing the type or produces types that are wider than possible. The `TypeIs` spec allows for some flexibility here, and type checkers are not obliged to always produce "optimal" (narrowest possible) type, so pyright is spec-compliant once this bug fix is in place. 

I consider this second part an enhancement request. Producing optimal types in this case is actually quite tricky because the type checker must effectively compute the intersection between two tuple types, and tuple types can be quite complicated. For example, how would one compute the (simplified) intersection of `tuple[str, *tuple[int | str, ...], T]` and `tuple[str | int, ...]`? For now, I'm going to reject this enhancement request, but I'm willing to reconsider this decision it in the future if there's sufficient demand from pyright users.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.374.
